sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcUtils.java

```diff
@@ -0,0 +1,111 @@
+                return FieldType.STRING;
+            case Types.BIT:
+            case Types.BOOLEAN:
+                return FieldType.BOOLEAN;
+            case Types.NUMERIC:
+                return FieldType.DECIMAL;
+            case Types.TINYINT:
+                return FieldType.BYTE;
+            case Types.SMALLINT:
+                return FieldType.INT16;
+            case Types.INTEGER:
+                return FieldType.INT32;
+            case Types.BIGINT:
+                return FieldType.INT64;
+            case Types.FLOAT:
+                return FieldType.FLOAT;
+            case Types.REAL:
+            case Types.DOUBLE:
+                return FieldType.DOUBLE;
+            case Types.VARBINARY:
+            case Types.BINARY:
+            case Types.BLOB:
+                return FieldType.BYTES;
+            case Types.DATE:
+                return FieldType.logicalType(
+                        new LogicalTypes.PassThroughLogicalType<Instant>(
+                                "SqlDateType", "", FieldType.DATETIME) {});
+            case Types.TIME:
+                return FieldType.logicalType(
+                        new LogicalTypes.PassThroughLogicalType<Instant>(
+                                "SqlTimeType", "", FieldType.DATETIME) {});
+            case Types.TIMESTAMP:
+                return FieldType.logicalType(
+                        new LogicalTypes.PassThroughLogicalType<Instant>(
+                                "SqlTimestampWithLocalTzType", "", FieldType.DATETIME) {});
+            default:
+                throw new UnsupportedOperationException( "Conversion or sql type to Beam type is unsupported");
+        }
+    }
+
+    public static Schema fromResultSetMetaDataToSchema(ResultSetMetaData metaData) throws SQLException  {
+        Schema.Builder schemaBuilder = Schema.builder();
+        for (int i = 1; i <= metaData.getColumnCount(); i++) {
+            FieldType fieldType = fromFieldType(metaData.getColumnType(i));
+            Field field = Field.of(metaData.getColumnName(i), fieldType).withNullable(metaData.isNullable(i) != ResultSetMetaData.columnNoNulls);
+            schemaBuilder.addField(field);
+        }
+        return schemaBuilder.build();
+    }
+
+    /** Convert a JDBC {@link ResultSet} to a Beam {@link Schema}. */
+    public static Schema fromResultSetToSchema(ResultSet resultSet) throws SQLException {
+        return fromResultSetMetaDataToSchema(resultSet.getMetaData());
+    }
+
+
+    /**
+     * Tries to convert a JDBC {@link ResultSet} row into a Beam {@link Row}.
+     */
+    public static Row toBeamRow(Schema rowSchema, ResultSet resultSet) {
+        return rowSchema.getFields().stream()
+                .map(field -> {
+                    try {
+                        return toBeamRowFieldValue(field, resultSet.getObject(field.getName()));
+                    } catch (SQLException e) {
+                        return null; // fixme: handle exception
```

Comment:
Returning null is probably not the right choice here because it will hide potential problems with the underlying database. The user could be fooled into thinking that the data is actually null and could end up calculating incorrect results.

Reply:
Done